### DISCUSSION TITLE
Intended color palette ?

### DISCUSSION THREAD

**Discussion by Samuel-DEVULDER** [spotlessmind1975/ugbasic#280] [General]
From the code, it seem the intended color palette is:
```
            { 0x00, 0x00, 0x00, 0, "BLACK" },
            { 0xff, 0xff, 0xff, 1, "WHITE" }, 
            { 0x88, 0x00, 0x00, 2, "RED" },
            { 0xaa, 0xff, 0xe6, 3, "CYAN" },
            { 0xcc, 0x44, 0xcc, 4, "VIOLET" },
            { 0x00, 0xcc, 0x55, 5, "GREEN" },
            { 0x00, 0x00, 0xaa, 6, "BLUE" },
            { 0xee, 0xee, 0x77, 7, "YELLOW" },
            { 0xa1, 0x68, 0x3c, 8, "ORANGE" },
            { 0xdd, 0x88, 0x65, 9, "BROWN" },
            { 0xff, 0x77, 0x77, 10, "LIGHT RED" },
            { 0x33, 0x33, 0x33, 11, "DARK GREY" },
            { 0x77, 0x77, 0x77, 12, "GREY" },
            { 0xaa, 0xff, 0x66, 13, "LIGHT GREEN" },
            { 0x00, 0x88, 0xff, 14, "LIGHT BLUE" },
            { 0xbb, 0xbb, 0xbb, 15, "LIGHT GREY" }
```
![image](https://user-images.githubusercontent.com/33961444/144650295-46fc4f82-f539-45be-a4d8-7d43bb8cefd8.png)
These seem to be somehow close to the [C64 color palette](http://unusedino.de/ec64/technical/misc/vic656x/colors/):
![image](https://user-images.githubusercontent.com/33961444/144650338-4e579453-9db3-4097-8b5b-2df9b89c58fc.png)
But they dont exactly match on the ORANGE/BROWN/LIGHT RED colors near the middle of the palette.

Beside this, there exists the [Dawnbringer-16](https://lospec.com/palette-list/dawnbringer-16) color palette which is designed to have good properties for pixel-art (see the discussion [here](https://pixeljoint.com/forum/forum_posts.asp?TID=12795)].
![image](https://user-images.githubusercontent.com/33961444/144652945-2d58f583-a97e-4ced-9ca0-0e2af1027275.png)

And as a matter of fact, it look to me that the dawnbringer palette comes between the default one for ugBasic and the C64. Is that an intended choice ?

**Comment by Samuel-DEVULDER** [comment]
**Side note**: The ef936x circuit has a pretty high gamma-value (around 2.8 from [the spec (page 6-87)](http://www.logicielsmoto.com/documents/ef9369.pdf)). This mean that it is not capable to produce low intensity r/g/b components (Actually the first non-zero level on the Thomson's matches level 100 out of 255 of RGB on PC). Or conversely that the simple linear mapping done in ef936x.c produces very bright colors and the pictures are kind of washed out. 

See the comparison between the palette on the PC128op and the intended palette.
![image](https://user-images.githubusercontent.com/33961444/144656225-681067f7-d435-41d0-b958-27d3df6d1d3a.png)
vs intended colors:
<img src="https://user-images.githubusercontent.com/33961444/144650295-46fc4f82-f539-45be-a4d8-7d43bb8cefd8.png" width="800">
We cannot really see any difference between brown and light-red on the PC128op, and the red on the left is way too bright in comparison to the intended color.

=> _So I'd suggest to introduce a bit of gamma in the conversion process to improve the contrast_ .

The value of gamma=2.8 is a bit too high since it maps 0x33 to 15*(0x33/255)^2.8 = 0.17 which is rounded up to 0. But a gamma of log(1/15)/log(0x33/255) = 1.68 will map 0x33 to ef936x's level 1 and $ff to level 15. This formula gives the following palette with the EF936x:

ugbc's palette on EF936x with gamma-correction
![image](https://user-images.githubusercontent.com/33961444/144655751-160394b5-e4b0-4cf3-ad90-9999a320781e.png)
vs intended palette:
<img src="https://user-images.githubusercontent.com/33961444/144650295-46fc4f82-f539-45be-a4d8-7d43bb8cefd8.png" width="800">
vs without-gamma-correction:
![image](https://user-images.githubusercontent.com/33961444/144656225-681067f7-d435-41d0-b958-27d3df6d1d3a.png)
Gamma-corrected colors now seem to better match intended ones :) 

An to go further here is the C64 palette converted with the gamma correction:
![image](https://user-images.githubusercontent.com/33961444/144655274-ee20dfd4-4fe6-4ec9-935d-4d9203f80bc4.png)
Intended:
<img src="https://user-images.githubusercontent.com/33961444/144650338-4e579453-9db3-4097-8b5b-2df9b89c58fc.png" width="800">
They are pretty close now.

And finally the "EF936x-gamma-corrected" Dawnbringer-16:
![image](https://user-images.githubusercontent.com/33961444/144659305-0febe8e1-f360-4340-b12c-1b023d0ae5bb.png)
vs real dawnbringer-16
<img src="https://user-images.githubusercontent.com/33961444/144652945-2d58f583-a97e-4ced-9ca0-0e2af1027275.png" width="800">
(still a bit too washed-out, but this because we didn't use a 2.8 gamma and dawnbringer requires low levels of colors the EF963x cannot produce and that needed to be emulated via [dithering](http://www.logicielsmoto.com/phpBB/viewtopic.php?p=5317&sid=fbd9d9a64f06d695be94776fdbab28eb#p5317))

In my own opinion the gamma formula ef963x = 15*(pc_value/255)^1.67 gives a pretty good approximated palette on the thomson, pretty close to the intended palette from the RGB values. The result look much better than the linear formula: ef963x = pc_value>>4.

If you want to do the comparison, here is the bit of code I changed in ef963x.c
```
/* converts a PC color to a thomson color for ef936 considering its very high gamma value */
static int pc_to_ef936x(int _pc_color) {
	// if(1) return (_pc_color>>4)&15;
	double pc_color = _pc_color/255.0;
	double ef936_color = 15*pow(pc_color, 1.67);
	return 0x0F & (int)(ef936_color + 0.5);
}

void ef936x_finalization( Environment * _environment ) {

    int i;

    outhead0("COMMONPALETTE");
    out0("      fdb ");

    RGBi * palette;

    if ( commonPalette ) {
        palette = commonPalette;
    } else {
        palette = SYSTEM_PALETTE;
    }

    for( i=0; i<16; ++i ) {
        out4("$%x%x%x%s", 
			 pc_to_ef936x(palette[i].blue),
			 pc_to_ef936x(palette[i].green),
			 pc_to_ef936x(palette[i].red),
			 i == 15 ? "" : ",");
    }
	outline0("");
}

```

**Comment by spotlessmind1975** [comment]
Hi @Samuel-DEVULDER, **these are all excellent considerations, thank you!**

I would like to read calmly all your observations, because in my opinion you have identified the solution to a problem that I still have regarding the conversion algorithms of the frames for animations ([take a look here](https://retroprogramming.iwashere.eu/iav)). It is a chapter of MIDRES library that, sooner or later, I will have to pick up again to bring it to ugBASIC.

However, you should also add the `COLOR` and `PALETTE` commands, which are new entries in version 1.6, to your considerations.

In practice, it is possible with these commands to assign a different shade with respect to the preset system palette, working on single indexes or on the overall palette at once. The function that converts the gradient from RGB to hardware is specialized for each chipset, and is found in correspondence with its implementation file (it is called `rgbConverterFunction`). That function is the ideal candidate, in my opinion, for the correction algorithm you describe. If you want to try it you can use the ugBASIC RGB (r, g, b) function.
